ckanext/unaids/actions.py

```diff
@@ -264,3 +264,61 @@ def package_create(next_action, context, data_dict):
 
 def time_ago_from_timestamp(context, data_dict):
     return t.h.time_ago_from_timestamp(data_dict.get('timestamp'))
+
+
+def dataset_lock(context, data_dict):
+    _check_access('dataset_lock', context, data_dict)
+    dataset_id = t.get_or_bust(data_dict, "id")
+    if not t.h.dataset_lockable(dataset_id):
+        raise t.ValidationError(t._("Datasets of this type cannot be locked."))
+    locked_name = t.config.get("ckanext.unaids.locked_release_name", "Locked")
+    context['bypass_read_only'] = True
+    t.get_action("package_patch")(
+        context,
+        {"id": dataset_id, "locked": True}
+    )
+    try:
+        t.get_action("dataset_version_create")(context, {
+            "dataset_id": dataset_id,
+            "name": locked_name,
+            "notes": t._("This dataset has been made read-only by a system administrator")
+        })
+    except t.ValidationError:
+        t.get_action("package_patch")(
+            context,
+            {"id": dataset_id, "locked": False}
+        )
+        raise t.ValidationError(t._(
+            "Couldn't lock dataset because a release named '{}' already "
+            "exists. Please rename or delete this release first. ".format(locked_name)
+        ))
+
+
+def dataset_unlock(context, data_dict):
+    _check_access('dataset_lock', context, data_dict)
+    dataset_id = t.get_or_bust(data_dict, "id")
+    if not t.h.dataset_lockable(dataset_id):
+        raise t.ValidationError(t._("Datasets of this type cannot be locked / unlocked."))
+    locked_name = t.config.get("ckanext.unaids.locked_release_name", "Locked")
+    context['bypass_read_only'] = True
+    context['ignore_auth'] = True
+    t.get_action("package_patch")(
+        context,
+        {"id": dataset_id, "locked": False}
+    )
+    try:
+        versions = t.get_action("dataset_version_list")(
+            context,
+            {"dataset_id": data_dict["id"]}
+        )
+        assert versions, "No versions exist"
+        latest_version = versions[-1]
+        assert latest_version["name"] == locked_name, f"Expected latest version '{locked_name}'"
+        t.get_action("version_delete")(context, {"version_id": latest_version['id']})
+    except Exception as e:
+        log.exception(f"Failed to delete version '{locked_name}'' "
```

Comment:
log.exception automatically adds whole stack trace in logs. No need to add `{e}` here.